### PR TITLE
added possibility to disable autocomplete in login form

### PR DESCRIPTION
**Pull request by Rayn0r**
Some companies require `autocomplete="off"` for input forms to comply with their security policy.

Disabling autocomplete is done by simply adding
`'login_form_autocomplete' => false,`
to the config.php file.

In case there is no appropriate entry, auto completion will remain enabled by default.

Signed-off-by: Rayn0r Andre.Weidemann@web.de

### CODE REVIEW

File: core/templates/login.php [hunk 1]
```diff
@@ -4,6 +4,8 @@
 script('core', 'merged-login');
 
 use OC\Core\Controller\LoginController;
+$config = \OC::$server->getConfig();
+$autocomplete = $config->getSystemValue('login_form_autocomplete', true);
```
[the remark this answers is not in the document]
Comment (Rayn0r) [reply core/templates/login.php]: I did that. But now some of the automated test fail...

**Comment by danielkesselberg** [comment]
Hey @Rayn0r,

thank you for your contribution :tada: Regarding the failing tests. At some point there are checks that one array is equals to another array. Because you added a new element its required to add the element to that checks.

1) https://github.com/nextcloud/server/blob/e0f9257be933ae8605424aaf5a08860cfd4359ff/tests/Core/Controller/LoginControllerTest.php#L201

You may add `login_form_autocomplete => 'off'` to the same array as `throttle_delay`.

2) https://github.com/nextcloud/server/blob/e0f9257be933ae8605424aaf5a08860cfd4359ff/tests/Core/Controller/LoginControllerTest.php#L225

Same as 1.

**Comment by Rayn0r** [comment]
@danielkesselberg Thank you for your help so far.
I did write a patch for "tests/Core/Controller/LoginControllerTest.php" using createMock.
This is something I have never done before, so it might be total rubbish...

**Comment by danielkesselberg** [comment]
It was a lot harder to get this to work than I thought in the first place :disappointed: 

1. https://github.com/nextcloud/server/blob/e0f9257be933ae8605424aaf5a08860cfd4359ff/tests/Core/Controller/LoginControllerTest.php#L340-L344

`getSystemValue` is called twice (1x with login_form_autocomplete, 1x with lost_password_link). We need to fake both calls to this method. Just replace the code above with the one below. This should do the trick :see_no_evil: 

```
$this->config
	->expects($this->exactly(2))
	->method('getSystemValue')
	->will($this->returnValueMap([
		['login_form_autocomplete', true, true],
		['lost_password_link', '', false],
	]));
```

2. https://github.com/nextcloud/server/blob/e0f9257be933ae8605424aaf5a08860cfd4359ff/tests/Core/Controller/LoginControllerTest.php#L257-L261

Same as 1.

File: tests/Core/Controller/LoginControllerTest.php [hunk 2]
```diff
@@ -338,10 +343,12 @@ public function testShowLoginFormForUserNamed0() {
 			->method('isLoggedIn')
 			->willReturn(false);
 		$this->config
-			->expects($this->once())
+			->expects($this->exactly(2))
 			->method('getSystemValue')
-			->with('lost_password_link')
-			->willReturn(false);
+			->will($this->returnValueMap([
+			['login_form_autocomplete', true, false],
```
Comment (danielkesselberg) [inline tests/Core/Controller/LoginControllerTest.php]: Change `false` to `true` than login_form_autocomplete will be on (that should be the default case)

File: tests/Core/Controller/LoginControllerTest.php [hunk 3]
```diff
@@ -255,10 +257,12 @@ public function testShowLoginFormWithPasswordResetOption($canChangePassword,
...
```
Comment (danielkesselberg) [inline tests/Core/Controller/LoginControllerTest.php]: Change `false` to `true` than login_form_autocomplete will be on (that should be the default case)

File: tests/Core/Controller/LoginControllerTest.php [hunk 4]
```diff
@@ -199,6 +199,7 @@ public function testShowLoginFormWithErrorsInSession() {
 				'alt_login' => [],
 				'resetPasswordLink' => null,
 				'throttle_delay' => 1000,
+				'login_form_autocomplete' => 'on',
```
Comment (danielkesselberg) [inline tests/Core/Controller/LoginControllerTest.php]: Change `on` to `off` (no stub for method here - getSystemValue will return null and if(null) results in off)

File: tests/Core/Controller/LoginControllerTest.php [hunk 5]
```diff
@@ -223,6 +224,7 @@ public function testShowLoginFormForFlowAuth() {
...
```
Comment (danielkesselberg) [inline tests/Core/Controller/LoginControllerTest.php]: Change `on` to `off` (no stub for method here - getSystemValue will return null and if(null) results in off)

**Comment by Rayn0r** [comment]
@danielkesselberg You are awesome!
All checks have passed. :-)
Let's see if the PR goes through now...

File: core/templates/login.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rayn0r) [reply core/templates/login.php]: Thanks to @danielkesselberg the code even passed the automated tests.
Please take a look at the code again.

**Comment by ChristophWurst** [comment]
> All checks have passed

:rocket:

**Comment by Rayn0r** [comment]
@ChristophWurst 
Since it took a lot of tweaking to get the automated tests done, there are a lot of commits in here.
Would it be ok to rebase the git commits, like I did here: https://github.com/Rayn0r/server/commits/rebase-test ?

**Comment by ChristophWurst** [comment]
> Since it took a lot of tweaking to get the automated tests done, there are a lot of commits in here.
> Would it be ok to rebase the git commits, like I did here:

Absolutely! Just rebase and squash them :wink:

**Comment by Rayn0r** [comment]
@ChristophWurst 
Hmm.. This time 2 checks (72 and 73) have failed, although the source code hasn't. And I don't quite understand the cause from check's log messages.
Would you care to take a short look?

**Comment by ChristophWurst** [comment]
https://drone.nextcloud.com/nextcloud/server/12054/279 looks unrelated
https://drone.nextcloud.com/nextcloud/server/12054/282 looks like it timed out

**Comment by rullzer** [comment]
Could you add the setting to `config.sample.php` with some documentation?

**Comment by Rayn0r** [comment]
> Could you add the setting to `config.sample.php` with some documentation?

I already prepared something [here](https://github.com/Rayn0r/documentation/commit/91b345f566b6ce34720738dfb304c6caccb2e762). 
I might as well add it to the config.sample.php

**Comment by danielkesselberg** [comment]
> configuration_server/config_sample_php_parameters.rst is auto-generated from the core config.sample.php file; changes to this file must be made in core https://github.com/nextcloud/server/tree/master/config

(from https://github.com/nextcloud/documentation#manuals).

**Review by rullzer — APPROVED** [review]
Lets do this :rocket:

**Review by ChristophWurst — APPROVED** [review]
Looks good

**Pull request merged**